### PR TITLE
Visibility timing

### PR DESCRIPTION
**Pull request by reshadi**
Fixed the timing of the component visibility events. Also added them to the graph

### CODE REVIEW

File: packages/hyperion-util/src/performanceAbsoluteNow.ts [hunk 1]
```diff
@@ -16,12 +16,17 @@
...
+type GetTimeFunc = () => number;
+type GetTimeFuncExtensions = {
+  setFallback: (fallback: () => number) => void;
+  fromRelativeTime(timestamp: number): number;
 };
 
+let performanceAbsoluteNow: GetTimeFunc & GetTimeFuncExtensions;
+//  {
+//   (): number,
+// } & ;
```
Comment (vsylee) [inline packages/hyperion-util/src/performanceAbsoluteNow.ts:28]: Delete?

File: packages/hyperion-autologging/src/ALSurfaceVisibilityPublisher.ts [hunk 2]
```diff
@@ -20,14 +20,24 @@ export type ALSurfaceVisibilityEventData =
...
     {
       surface: string;
-      event: 'component_visibility';
-      isIntersecting: boolean;
       intersectionEntry: IntersectionObserverEntry;
-    }
+    } &
+    (
+      {
+        event: 'component_visibile';
```
Comment (vsylee) [inline packages/hyperion-autologging/src/ALSurfaceVisibilityPublisher.ts:32]:
```suggestion
        event: 'component_visible';
```

File: packages/hyperion-autologging/src/ALSurfaceVisibilityPublisher.ts [hunk 3]
```diff
@@ -155,19 +165,26 @@ export function publish(options: InitOptions): void {
                 console.warn("Don't know yet how to merge entries!");
               }
 
+              const isIntersecting = entry.isIntersecting;
               channel.emit('al_surface_visibility_event', {
-                event: 'component_visibility',
-                eventTimestamp: performanceAbsoluteNow(),
+                ...isIntersecting
+                  ? { event: 'component_visibile', isIntersecting }
```
Comment (vsylee) [inline packages/hyperion-autologging/src/ALSurfaceVisibilityPublisher.ts:171]:
```suggestion
                  ? { event: 'component_visible', isIntersecting }
```

File: packages/hyperion-autologging/src/ALSurfaceVisibilityPublisher.ts [hunk 4]
```diff
@@ -155,19 +165,26 @@ export function publish(options: InitOptions): void {
...
+                  : { event: 'component_hidden', isIntersecting },
+                // event: 'component_visibility',
```
Comment (vsylee) [inline packages/hyperion-autologging/src/ALSurfaceVisibilityPublisher.ts:173]: Delete

File: packages/hyperion-autologging/src/ALSurfaceVisibilityPublisher.ts [hunk 5]
```diff
@@ -155,19 +165,26 @@ export function publish(options: InitOptions): void {
...
                 element: surfaceEvent.element,
                 autoLoggingID: surfaceEvent.autoLoggingID, // same element, same ID
-                metadata: {},
+                metadata: {
+                  emit_time: '' + performanceAbsoluteNow(), // just to keep track of the difference
+                },
                 callFlowlet: surfaceEvent.callFlowlet,
                 triggerFlowlet: surfaceEvent.triggerFlowlet,
-                isIntersecting: entry.isIntersecting,
+                // isIntersecting,
```
Comment (vsylee) [inline packages/hyperion-autologging/src/ALSurfaceVisibilityPublisher.ts:185]: Delete

File: packages/hyperion-util/src/performanceAbsoluteNow.ts [hunk 6]
```diff
@@ -38,15 +43,24 @@ if (performanceIsDefined) {
...
 } else {
-  performanceAbsoluteNow = () => fallback();
+  coreFunction = () => fallback();
 }
-performanceAbsoluteNow.setFallback = setFallback;
+
+const extensions: GetTimeFuncExtensions = {
+  setFallback,
+  fromRelativeTime: (timestamp: number) => {
+    return timestamp + navigationStart;
```
Comment (vsylee) [inline packages/hyperion-util/src/performanceAbsoluteNow.ts:60]: Could there be any unexpected offsets if `navigationStart` is still -1 at this point?

**Review by vsylee — APPROVED** [review]

File: packages/hyperion-util/src/performanceAbsoluteNow.ts (hunk 6 above)
Comment (reshadi) [reply packages/hyperion-util/src/performanceAbsoluteNow.ts:60]: I debated with myself on this one too. If `navigationStart` is still -1 at this point, then we will fallback to the Date.now and it means that the application also does not have access to `performance` and hence cannot pass relative values to the function. 
So, in practice, it should not happen. I add a warning to at least ensure this is not happening.

**Pull request merged**